src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,634 @@
+	}
+}
+
+/**
+* Poll for received packets
+*
+* @v netdev		Network device
+*/
+void atl_poll_rx(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+	struct atl_desc_rx_wb *rx;
+	struct io_buffer *iobuf;
+	size_t len;
+
+	/* Check for received packets */
+	while (nic->rx_ring.sw_head != nic->rx_ring.sw_tail) {
+
+		/* Get next receive descriptor */
+		rx = (struct atl_desc_rx_wb *)nic->rx_ring.ring +
+		nic->rx_ring.sw_head;
+
+		/* Stop if descriptor is still in use */
+		DBG("AQUANTIA: rx poll: desc: %llx, %llx\n", *((uint64_t *)rx),
+		    *(((uint64_t *)rx) + 1));
+		if (!rx->dd)
+			return;
+
+		/* Populate I/O buffer */
+		iobuf = nic->iobufs[nic->rx_ring.sw_head];
+		nic->iobufs[nic->rx_ring.sw_head] = NULL;
+		len = le16_to_cpu(rx->pkt_len);
+		iob_put(iobuf, len);
+
+		/* Hand off to network stack */
+		/*to do: process error*/
+		DBG("AQUANTIA: %p RX[%d] complete (length %zd)\n",
+		    nic, nic->rx_ring.sw_head, len);
+		netdev_rx(netdev, iobuf);
+		DBG("AQUANTIA %p: RX[%d] complete\n",
+		    nic, nic->rx_ring.sw_head);
+		atl_ring_next_dx(&nic->rx_ring.sw_head);
+	}
+}
+
+/**
+* Poll for completed and received packets
+*
+* @v netdev		Network device
+*/
+static void atl_poll(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+
+	/* Check link state */
+	atl_check_link(netdev);
+
+	/* Poll for TX completions, if applicable */
+	/* if (icr & ATL_IRQ_TX) */
+	atl_poll_tx(netdev);
+
+	/* Poll for RX completions, if applicable */
+	/* if (icr & ATL_IRQ_RX) */
+	atl_poll_rx(netdev);
+
+	/* Refill RX ring */
+	atl_rx_ring_fill(nic);
+}
+
+/**
+* Enable or disable interrupts
+*
+* @v netdev		Network device
+* @v enable		Interrupts should be enabled
+*/
+static void atl_irq(struct net_device *netdev, int enable)
+{
+	struct atl_nic *nic = netdev->priv;
+	uint32_t mask;
+
+	DBG("AQUANTIA: irq: %d\n", enable);
+
+	mask = (ATL_IRQ_TX | ATL_IRQ_RX);
+	if (enable)
+		ATL_WRITE_REG(mask, ATL_ITR_MSKS);
+	else
+		ATL_WRITE_REG(mask, ATL_ITR_MSKC);
+
+}
+
+/** Marvell network device operations */
+static struct net_device_operations atl_operations = {
+	.open = atl_open,
+	.close = atl_close,
+	.transmit = atl_transmit,
+	.poll = atl_poll,
+	.irq = atl_irq,
+};
+
+/******************************************************************************
+*
+* PCI interface
+*
+*******************************************************************************
+*/
+
+/**
+* Probe PCI device
+*
+* @v pci		PCI device
+* @ret rc		Return status code
+*/
+static int atl_probe(struct pci_device *pci)
+{
+	struct net_device *netdev;
+	struct atl_nic *nic;
+	int rc = ENOERR;
+
+	DBG("\nAQUANTIA: atl_probe()\n");
+
+	/* Allocate and initialize net device */
+	netdev = alloc_etherdev(sizeof(*nic));
+	if (!netdev) {
+		rc = -ENOMEM;
+		goto err_alloc;
+	}
+	netdev_init(netdev, &atl_operations);
+	nic = netdev->priv;
+	pci_set_drvdata(pci, netdev);
+	netdev->dev = &pci->dev;
+	memset(nic, 0, sizeof(*nic));
+	nic->port = PCI_FUNC(pci->busdevfn);
```

Comment:
How will this behave when using PCI pass-through inside a VM?  For passed-through devices, the virtual PCI function number is likely to be zero regardless of the real function number.  See e.g. commit https://github.com/ipxe/ipxe/commit/76a7bfe93